src/draft06.rs

```diff
@@ -0,0 +1,756 @@
+) -> Result<(AeadKey, AeadNonce)> {
+    let buf = compose(query)?;
+    let salt = [
+        buf.as_ref(),
+        &to_u16(response_nonce.len())?.to_be_bytes(),
+        &response_nonce,
+    ]
+    .concat();
+
+    let h_key = Hkdf::<<Kdf as KdfTrait>::HashImpl>::new(Some(&salt), &odoh_secret);
+    let mut key = AeadKey::default();
+    h_key
+        .expand(LABEL_KEY, &mut key)
+        .map_err(|_| Error::from(HpkeError::InvalidKdfLength))?;
+
+    let h_nonce = Hkdf::<<Kdf as KdfTrait>::HashImpl>::new(Some(&salt), &odoh_secret);
+    let mut nonce = AeadNonce::default();
+    h_nonce
+        .expand(LABEL_NONCE, &mut nonce)
+        .map_err(|_| Error::from(HpkeError::InvalidKdfLength))?;
+
+    Ok((key, nonce))
+}
+
+#[inline]
+fn to_u16(n: usize) -> Result<u16> {
+    n.try_into().map_err(|_| Error::InvalidInputLength)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use rand::rngs::StdRng;
+    use rand::SeedableRng;
+
+    #[test]
+    fn configs() {
+        // parse
+        let configs_hex = "002cff0600280020000100010020bbd80565312cff62c44020a60c511711a6754425d5f42be1de3bca6b9bb3c50f";
+        let mut configs_bin: Bytes = hex::decode(configs_hex).unwrap().into();
+        let configs: ObliviousDoHConfigs = parse(&mut configs_bin).unwrap();
+        assert_eq!(configs.configs.len(), 1);
+        // check all bytes have been consumed
+        assert!(configs_bin.is_empty());
+
+        // compose
+        let buf = compose(&configs).unwrap();
+        assert_eq!(configs_hex, hex::encode(&buf));
+    }
+
+    #[test]
+    fn pubkey() {
+        // parse
+        let key_hex =
+            "0020000100010020aacc53b3df0c6eb2d7d5ce4ddf399593376c9903ba6a52a52c3a2340f97bb764";
+        let mut key_bin: Bytes = hex::decode(key_hex).unwrap().into();
+        let key: ObliviousDoHConfigContents = parse(&mut key_bin).unwrap();
+        assert!(key_bin.is_empty());
+
+        // compose
+        let buf = compose(&key).unwrap();
+        assert_eq!(key_hex, hex::encode(&buf));
+    }
+
+    #[test]
+    fn exchange() {
+        // Use a seed to initialize a RNG. *Note* you should rely on some
+        // random source.
+        let mut rng = StdRng::from_seed([0; 32]);
+
+        // Generate a key pair on server side.
+        let key_pair = ObliviousDoHKeyPair::new(&mut rng);
+
+        // Create client configs from the key pair. It can be distributed
+        // to the clients.
+        let public_key = key_pair.public().clone();
+        let client_configs: ObliviousDoHConfigs = vec![ObliviousDoHConfig::from(public_key)].into();
+        let client_configs_bytes = compose(&client_configs).unwrap().freeze();
+
+        // ... distributing client_configs_bytes ...
+
+        // Parse and extract first supported config from client configs on client side.
+        let client_configs: ObliviousDoHConfigs = parse(&mut client_configs_bytes.clone()).unwrap();
+        let client_config = client_configs.into_iter().next().unwrap();
```

Comment:
The client config should not be manually extracted, this should be done by the library. See example here: https://github.com/cloudflare/odoh-client-rs/blob/master/src/main.rs#L52